internal/services/containerapps/helpers/container_apps.go

```diff
@@ -0,0 +1,1803 @@
+		probes = append(probes, expandContainerAppLivenessProbe(input.LivenessProbe[0]))
+	}
+
+	if len(input.ReadinessProbe) == 1 {
+		probes = append(probes, expandContainerAppReadinessProbe(input.ReadinessProbe[0]))
+	}
+
+	if len(input.StartupProbe) == 1 {
+		probes = append(probes, expandContainerAppStartupProbe(input.StartupProbe[0]))
+	}
+
+	return &probes
+}
+
+type Secret struct {
+	Name  string `tfschema:"name"`
+	Value string `tfschema:"value"`
+}
+
+func SecretsSchema() *pluginsdk.Schema {
+	return &pluginsdk.Schema{
+		Type:         pluginsdk.TypeSet,
+		Optional:     true,
+		Sensitive:    true,
+		ValidateFunc: nil,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
+				"name": {
+					Type:         pluginsdk.TypeString,
+					Required:     true,
+					ValidateFunc: validate.SecretName,
+					Sensitive:    true,
+					Description:  "The Secret name.",
+				},
+
+				"value": {
+					Type:        pluginsdk.TypeString,
+					Required:    true,
+					Sensitive:   true,
+					Description: "The value for this secret.",
+				},
+			},
+		},
+	}
+}
+
+func SecretsDataSourceSchema() *pluginsdk.Schema {
+	return &pluginsdk.Schema{
+		Type:      pluginsdk.TypeList,
+		Computed:  true,
+		Sensitive: true,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
+				"name": {
+					Type:        pluginsdk.TypeString,
+					Computed:    true,
+					Description: "The Secret name.",
+				},
+
+				"value": {
+					Type:        pluginsdk.TypeString,
+					Computed:    true,
+					Sensitive:   true,
+					Description: "The value for this secret.",
+				},
+			},
+		},
+	}
+}
+
+func ExpandContainerSecrets(input []Secret) *[]containerapps.Secret {
+	if len(input) == 0 {
+		return nil
+	}
+
+	result := make([]containerapps.Secret, 0)
+
+	for _, v := range input {
+		result = append(result, containerapps.Secret{
+			Name:  pointer.To(v.Name),
+			Value: pointer.To(v.Value),
+		})
+	}
+
+	return &result
+}
+
+func ExpandFormerContainerSecrets(metadata sdk.ResourceMetaData) *[]containerapps.Secret {
+	secretsRaw, _ := metadata.ResourceData.GetChange("secret")
+	result := make([]containerapps.Secret, 0)
+	if secrets, ok := secretsRaw.([]interface{}); ok {
+
+		for _, secret := range secrets {
+			if v, ok := secret.(map[string]interface{}); ok {
+				result = append(result, containerapps.Secret{
+					Name:  pointer.To(v["name"].(string)),
+					Value: pointer.To(v["value"].(string)),
+				})
+			}
+		}
+	}
+
+	return &result
+}
+
+func UnpackContainerSecretsCollection(input *containerapps.SecretsCollection) *[]containerapps.Secret {
+	if input == nil || len(input.Value) == 0 {
+		return nil
+	}
+
+	result := make([]containerapps.Secret, 0)
+	for _, v := range input.Value {
+		result = append(result, containerapps.Secret(v))
+	}
+
+	return &result
+}
+
+func UnpackContainerDaprSecretsCollection(input *daprcomponents.DaprSecretsCollection) *[]daprcomponents.Secret {
+	if input == nil || len(input.Value) == 0 {
+		return nil
```

Comment:
presumably this should be returning an empty slice rather than nil?

Reply:
as above